### ISSUE TITLE
Envelope Scheduling Issue

### ISSUE THREAD

**Issue by dveca** [WebAudio/web-audio-api#510]
The code examples below illustrates a problem we are having implementing amplitude 
enveloping (e.g., ADSR).  For simplicity, only Attack and Decay are implemented 
in the code below.

This example works fine at slow tempos (e.g., quarter notes at 60bpm), but at 
faster rates (e.g., quarter notes at 400bpm or thirty-second notes at 60bpm)
a problem arises due to the fact that after a note_on is issued, the 
attack ramp hasn't fully completed before the note_off is issued, resulting
in gain discontinuity (a "pop"), or requiring an undesirable workaround 
to prevent it.

The start() method works fine, is included for reference; however, the stop() method is problematic
as shown below.

Osc.prototype.start = function(start_time)
{ 
   var zero = 0.0000001;
    
    // Init Gain Node.
    this.gain_node = this.audio_context.createGain();
    this.gain_node.gain.setValueAtTime(zero, this.audio_context.currentTime);
    this.gain_node.connect(this.audio_context.destination);
    
    // Init Osc Node.
    this.osc_node = this.audio_context.createOscillator();
    this.osc_node.type  = this.wave_type;
    this.osc_node.frequency.value = this.frq;
    this.osc_node.detune.value = this.detune;
    this.osc_node.connect(this.gain_node);
    
    // Start Osc Node.
    this.osc_node.start(start_time);
    this.gain_node.gain.setValueAtTime(zero, start_time);
    this.gain_node.gain.exponentialRampToValueAtTime(this.amp * this.gain, start_time + this.attack);
};

//	Solution 1
//	The problem here is that if the attack is still ramping up, "this.gain_node.gain.value"
//	is not the correct starting gain for the ramp down to zero, and the result is sudden gain 
//	discontinuity resulting in an audible "pop."
Osc.prototype.stop = function(stop_time)
{    
    this.gain_node.gain.cancelScheduledValues(stop_time);
    this.gain_node.gain.setValueAtTime(this.gain_node.gain.value, stop_time);
    this.gain_node.gain.exponentialRampToValueAtTime(0.000001, stop_time + this.decay);
    this.osc_node.stop(stop_time + this.decay); 
};

//	Solution 2
//	While this solution avoids the "pop," the gain does not reach the level it would have reached
//	had the ramp-up continued until "stop_time."  This solution isn't the desired effect, and 
//	is unfortunately the best a work-around.
Osc.prototype.stop = function(stop_time)
{    
    this.gain_node.gain.cancelScheduledValues(this.audio_context.currentTime);
    this.gain_node.gain.setValueAtTime(this.gain_node.gain.value, this.audio_context.currentTime);
    this.gain_node.gain.exponentialRampToValueAtTime(0.000001, stop_time + this.decay);
    this.osc_node.stop(stop_time + this.decay); 
};

//	Solution 3
//	The problem with this solution is that the third param to setTargetAtTime() is not a time parameter.  
//	If it were, then I could use this.decay, and the envelope would fade correctly.
Osc.prototype.stop = function(stop_time)
{    
   var time_constant = 0.5; // Arbitrary value.
   this.gain_node.gain.setTargetAtTime(0.000001, stop_time, time_constant);
   this.osc_node.stop(stop_time + this.decay); 
};

//	Preferred Solution
//	It seems we would need something similar to the following. If the third param to setTargetAtTime()
//	were in fact a time value, then the envelope would fade correctly.
Osc.prototype.stop = function(stop_time)
{    
   var decay_time = this.decay;
   this.gain_node.gain.setTargetAtTime(0.000001, stop_time, decay_time);
   this.osc_node.stop(stop_time + this.decay);
};

Please also see "illustration.png" for a visual description of the problem.
![illustration](https://cloud.githubusercontent.com/assets/11874967/7072867/3cbbaa9e-dea4-11e4-8c8b-6847c04738fd.png)

**Comment by mark-buer** [comment]
Perhaps AudioParam could be augmented with a new method:
```c
void holdValueAtTime (float value, double startTime);
```
or this method, which seems more simple/flexible:
```c
float getValueAtTime (double startTime);
```
?

@dveca As a workaround, could you calculate what the value would be at stop_time by applying the [equations given in the spec](http://webaudio.github.io/web-audio-api/#widl-AudioParam-exponentialRampToValueAtTime-void-float-value-double-endTime) due to the start envelope?

> The value during the time interval T0 <= t < T1 (where T0 is the time of the previous event and T1 is the endTime parameter passed into this method) will be calculated as:

>        v(t) = V0 * (V1 / V0) ^ ((t - T0) / (T1 - T0))

In the example you've given (just a start envelope) the calculations seem simple enough, but for more complex scenarios you'll essentially need to duplicate the AudioParam code from within a WebAudio implementation ([AudioParam](https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/modules/webaudio/AudioParam.cpp) and [AudioParamTimeline](https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/modules/webaudio/AudioParamTimeline.cpp) from Chromium for example).

**Comment by rtoy** [comment]
This doesn't address the stated issue, but aren't attack and decay usually
done with an exponential approach available as setTargetAtTime?  For just
attack and decay, the setTargetAtTime seems to do the right thing.  The
decay will start at whatever the attack value is at the point where the
decay begins.

But I can see mark-buer's idea for a holdValueAtTime(time) is useful.  This
would set the value to be whatever the in-progress automation value would
be at time |time|, effectively stopping any automation at that time, and
holding the value.

I didn't quite follow why mark-buer wants holdValueAtTime(value,
startTime).  What is |value| here?

On Thu, Apr 9, 2015 at 10:39 AM, dveca <notifications@github.com> wrote:

> The code examples below illustrates a problem we are having implementing
> amplitude
> enveloping (e.g., ADSR). For simplicity, only Attack and Decay are
> implemented
> in the code below.
>
> This example works fine at slow tempos (e.g., quarter notes at 60bpm), but
> at
> faster rates (e.g., quarter notes at 400bpm or thirty-second notes at
> 60bpm)
> a problem arises due to the fact that after a note_on is issued, the
> attack ramp hasn't fully completed before the note_off is issued, resulting
> in gain discontinuity (a "pop"), or requiring an undesirable workaround
> to prevent it.
>
> The start() method works fine, is included for reference; however, the
> stop() method is problematic
> as shown below.
>
> Osc.prototype.start = function(start_time)
> {
> var zero = 0.0000001;
>
> // Init Gain Node.
> this.gain_node = this.audio_context.createGain();
> this.gain_node.gain.cancelScheduledValues(this.audio_context.currentTime);  // Voodoo.
> this.gain_node.gain.setValueAtTime(zero, this.audio_context.currentTime);
> this.gain_node.connect(this.audio_context.destination);
> if (IsDefined(this.delay_send_gain_node))
> {
>     this.gain_node.connect(this.delay_send_gain_node);
> }
>
> // Init Osc Node.
> this.osc_node                   = this.audio_context.createOscillator();
> this.osc_node.type              = this.wave_type;
> this.osc_node.frequency.value   = this.frq;
> this.osc_node.detune.value      = this.detune;
> this.osc_node.connect(this.gain_node);
>
> // Start Osc Node.
> this.osc_node.start(start_time);
> this.gain_node.gain.setValueAtTime(zero, start_time);
> this.gain_node.gain.exponentialRampToValueAtTime(this.amp * this.gain, start_time + this.attack);
>
> };
>
> // Solution 1
> // The problem here is that if the attack is still ramping up,
> "this.gain_node.gain.value"
> // is not the correct starting gain for the ramp down to zero, and the
> result is sudden gain
> // discontinuity resulting in an audible "pop."
> Osc.prototype.stop = function(stop_time)
> {
>
> this.gain_node.gain.cancelScheduledValues(stop_time);
> this.gain_node.gain.setValueAtTime(this.gain_node.gain.value, stop_time);
> this.gain_node.gain.exponentialRampToValueAtTime(0.000001, stop_time +
> this.decay);
> this.osc_node.stop(stop_time + this.decay);
> };
>
> // Solution 2
> // While this solution avoids the "pop," the gain does not reach the level
> it would have reached
> // had the ramp-up continued until "stop_time." This solution isn't the
> desired effect, and
> // is unfortunately the best a work-around.
> Osc.prototype.stop = function(stop_time)
> {
>
> this.gain_node.gain.cancelScheduledValues(this.audio_context.currentTime);
> this.gain_node.gain.setValueAtTime(this.gain_node.gain.value,
> this.audio_context.currentTime);
> this.gain_node.gain.exponentialRampToValueAtTime(0.000001, stop_time +
> this.decay);
> this.osc_node.stop(stop_time + this.decay);
> };
>
> // Solution 3
> // The problem with this solution is that the third param to
> setTargetAtTime() is not a time parameter.
>
> // If it were, then I could use this.decay, and the envelope would fade
> correctly.
> Osc.prototype.stop = function(stop_time)
> {
>
> var time_constant = 0.5; // Arbitrary value.
> this.gain_node.gain.setTargetAtTime(0.000001, stop_time, time_constant);
> this.osc_node.stop(stop_time + this.decay);
> };
>
> // Preferred Solution
> // It seems we would need something similar to the following. If the third
> param to setTargetAtTime()
> // were in fact a time value, then the envelope would fade correctly.
> Osc.prototype.stop = function(stop_time)
> {
>
> var decay_time = this.decay;
> this.gain_node.gain.setTargetAtTime(0.000001, stop_time, decay_time);
> this.osc_node.stop(stop_time + this.decay);
> };
>
> Please also see "illustration.png" for a visual description of the problem.
> [image: illustration]
> <https://cloud.githubusercontent.com/assets/11874967/7072867/3cbbaa9e-dea4-11e4-8c8b-6847c04738fd.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAudio/web-audio-api/issues/510>.
>



-- 
Ray

**Comment by mark-buer** [comment]
|value| is a failure to mash keys on keyboard correctly. Fixed.

**Comment by dveca** [comment]
I did in fact try setTargetAtTime().  The problem here is that the third param to setTargetAtTime() is not a time parameter; if it were, then I could use this.decay, and the envelope would fade correctly.  It certainly "works," but it doesn't produce the desired envelope decay-time result.  (See "Solution 3" in my original report.)

Mark actually suggested a possible work-around where I could calculate what the time would be at "stop_time," but that seems a bit clumsy and inefficient (especially calculating it from JS).  It would be much cooler to have it as part of the API.

Mark's workaround suggestion:
"As a workaround, could you calculate what the value would be at stop_time by applying the equations given in the spec<https://urldefense.proofpoint.com/v1/url?u=http://webaudio.github.io/web-audio-api/%23widl-AudioParam-exponentialRampToValueAtTime-void-float-value-double-endTime&k=g%2Fpg0ANArHUUOAU48pcA%2FA%3D%3D%0A&r=mfEh%2BTEltWj5l8cO2WOLPR7FTE7XRTVwjhqIA28GQjI%3D%0A&m=3PJaKav%2FNRdMHP5EeHqjuizGcBxPkKt95hy4ouUhv5o%3D%0A&s=1e226557e5d90e898f412ef4a09eb281dd5557bf4ac1f3ab961a6fc2f192552d> due to the start envelope?

The value during the time interval T0 <= t < T1 (where T0 is the time of the previous event and T1 is the endTime parameter passed into this method) will be calculated as:

   v(t) = V0 * (V1 / V0) ^ ((t - T0) / (T1 - T0))"

BTW, thanks, all, for getting back so soon on this!
dv

**Comment by dveca** [comment]
Mark, not sure what you mean by "|value| is a failure to mash keys on keyboard correctly."

**Comment by notthetup** [comment]
Interesting discussion. I just recently had to deal with this for some cross-fading. 

I am always afraid of "calculate what the value would be at stop_time by applying the equations given in the spec". I looked at doing that looking at #296, but I found the reimplementation in JS was never perfectly accurate as the native implementation, and the various browser implementations have minor differences which we have to somehow emulate in JS. Not something that's sustainable in the long run. 

What could be interesting here from a spec side of thing is to return a k-rate value in the getter of `this.gain_node.gain` instead of some form of "last stable value" which it does right now. But maybe this is related to #128 somehow as well.

**Comment by si-robertson** [comment]
I don't believe this issue is directly related to automated parameter values. Popping typically occurs in audio when a high sample amplitude is immediately followed a low sample amplitude, i.e. when the audio is stopped without any fade-out; the fade-out only needs to be a few samples long to avoid the problem. I have ran into this issue numerous times when using Web Audio.

Ideally, handling this issue would be spec'd and all browser implementations would deal with it behind the scenes. All they would need to do is repeat the last sample a few times while reducing its amplitude, and the sample window would be so small no one would notice it happening.

**Comment by notthetup** [comment]
OK. Fair enough. Even if the `.value` getters helped to figure out what the current (not sample accurate) value of the automated parameter was, one would still have to calculate the value at the intended start time for the release section. 

The parameter automation methods use some form of 'last stable' value to figure out the starting value for the automation. 

That's `V0` in this equation:

```
 v(t) = V0 * (V1 / V0) ^ ((t - T0) / (T1 - T0))
```

In your case, you need the V0 to be the current value of the gain at `stop_time`, but WebAudio takes it to be the 'last stable' value of the gain which was before your linearRamp automation started in the `.start` method.

I'm guessing (and I hope someone can correct me) that this is done due to the fact that the APIs like `linearRampToValueAtTime` don't have a start time parameter, so they have to start at the last known sample-accurate timestamp of something that happened due to a previous automation call.

The other solution to this would be 'userland' libraries which can track all the automation you have put in place and predict future values accurately.

**Comment by karlt** [comment]
On Thu, 09 Apr 2015 15:35:34 -0700, mark-buer wrote:

> Perhaps AudioParam could be augmented with a new method:
> ```c
> void holdValueAtTime (float value, double startTime);
> ```

setTargetAtTime() is the existing method for indicating an event
that does not immediately change the intrinsic value.

If the third parameter to setTargetAtTime could accept
unrestricted double then this would be equivalent.

```c
setTargetAtTime(unusedTarget, startTime, inf)
```

In the use case presented here, the value of the third parameter
could actually be unnecessary, but which equation should be used
between a setTargetAtTime event and a
linearRampToValueAtTime/exponentialRampToValueAtTime event is
inconsistently specified.

If only the linearRampToValueAtTime/exponentialRampToValueAtTime
curve were specified to apply after a setTargetAtTime event, then
the startTime passed to setTargetAtTime would indicate the start
time of the *RampToValueAtTime curve.

(I don't know of a situation where having a setTargetAtTime
curve apply before a *RampToValueAtTime event would be useful.)

**Comment by notthetup** [comment]
I found this issue bothering me today was while fading some AudioBufferSources today.

So basically the fade-in/out is triggered by user interaction and if it happens quickly enough, the fade-out starts before the fade-in has completed. Like what @dveca mentioned since there is no way to know the value of gain at the instance when the user triggered the fade-out, there is no way to make the fade-in/out smooth.

Here is a simple example with the various things we can add and how it sounds.

http://jsfiddle.net/notthetup/zrLb0pcy/

The only way I managed to make a specific type of smooth fade in/out work for me (since I'm only using `linearRampToValueAtTime`) by calculating the current value of the gain in the last example in the above JSFiddle

**Comment by joeberkovitz** [comment]
Closing since actionable changes to API here (calculating value at some point in automation schedule) is captured elsewhere..  Will file issue for clarification of cancelScheduledValues() effect on automation and on current value.

**Comment by mark-buer** [comment]
Reading the spec, this still appears to be an unresolved problem.

> Closing since actionable changes to API here (calculating value at some point in automation schedule) is captured elsewhere

Where is it captured?

> Will file issue for clarification of cancelScheduledValues() effect on automation and on current value.

Has this been done?

**Comment by joeberkovitz** [comment]
The former point is embedded in #573 and I believe that the latter point is addressed via #344 (which still requires some discussion).
